core/src/main/scala/kafka/coordinator/TransactionCoordinator.scala

```diff
@@ -0,0 +1,138 @@
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+package kafka.coordinator
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.server.{KafkaConfig, ReplicaManager}
+import kafka.utils.{Logging, Pool, ZkUtils}
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.utils.Time
+
+/**
+  * Transaction coordinator handles message transactions sent by producers and communicate with brokers
+  * to update ongoing transaction's status.
+  *
+  * Each Kafka server instantiates a transaction coordinator which is responsible for a set of
+  * producers. Producers with specific transactional ids are assigned to their corresponding coordinators;
+  * Producers with no specific transactional id may talk to a random broker as their coordinators.
+  */
+object TransactionCoordinator {
+
+  def apply(config: KafkaConfig, zkUtils: ZkUtils, time: Time): TransactionCoordinator = {
+
+    val pIDManager = new PIDManager(config.brokerId, zkUtils)
+    val logManager = new TransactionLogManager(config.brokerId, zkUtils)
+    new TransactionCoordinator(config.brokerId, pIDManager, logManager)
+  }
+}
+
+class TransactionCoordinator(val brokerId: Int,
+                             val pIDManager: PIDManager,
+                             val logManager: TransactionLogManager) extends Logging {
+
+  this.logIdent = "[Transaction Coordinator " + brokerId + "]: "
+
+  type InitPIDCallback = InitPIDResult => Unit
+
+  /* Active flag of the coordinator */
+  private val isActive = new AtomicBoolean(false)
+
+  /* TxnID to PID metadata map cache */
+  private val pIDMetadataCache = new Pool[String, PIDMetadata]
+
+  def handleInitPID(txnID: String,
+                    responseCallback: InitPIDCallback): Unit = {
+    if (txnID == null || txnID.isEmpty) {
+      // if the txnID is not specified, then always blindly accept the request
+      // and return a new PID from the PID manager
+      val pID: Long = pIDManager.getNewPID()
+
+      responseCallback(InitPIDResult(pID, 0 /* epoch */, Errors.NONE))
+    } else if(!logManager.isCoordinatorFor(txnID)) {
+      // check if it is the assigned coordinator for the txnID
+      responseCallback(initPIDError(Errors.NOT_COORDINATOR_FOR_GROUP))
```

Comment:
Seems we should rename this error code (not necessarily in this patch). Maybe just `NOT_COORDINATOR`?